### ISSUE TITLE
stackalloc[type](5)  desperately needs an implicit or new declaration with Int argument.

### ISSUE THREAD

**Issue by LeeTibbert** [scala-native/scala-native#2786]
#### TL;DR

I seek permission to add an implicit to convert `Int` to `USize`.  Such would add conceptual
clarity and make stackalloc & alloc much easier to use & review.

`USize.scala` already has an implicit for UInt
```
USize.scala:  @inline implicit def uintToUSize(x: UInt): USize = x.toUSize
```
My goal is to be able to write the  straight forward, relatively simple &
understandable   `stackalloc[byte](5)` that is documented and has been for a long time.
No type faffing.

Please advise.

#### Discussion
This discussion is related to Issue #2728.   I have been developing using the
code base with 32 bit support for two months now.  Overall, I am surprised and
gratified at how little disruption  that has been in 64 bit land and the ease of
the transition.  I have not use the 32 bit code and it has not used me, so I am
happy there.  Kudos Shadaj!

The one continuing place where my work flow gets interrupted and my tranquility
challenged is with the new type in stackalloc/alloc.  This hoses me about every
half hour, I am a slow learner, frequent forgetter, and dominated by past
experience.  Each time I encounter it, I add to my karmic debt.

In addition to being a personal pain, I suspect that it is and will be a pain point
for other developers.  

I also believe that there is also an argument from theory for fixing it, see below. 

##### Proposed fix
The fix is to add an implicit from `Int` to (a never actualized) `Count`
and one from `Count` to `USize`.   One can then write
`stackalloc[Byte](5)`  with consistent typing. 

##### Rationale

This history is reminds me of an M.C. Escher graphic which
starts off as a recognizable object on one side and then,
through a series of slight intermediate changes, ends up
as something entirely unrelated on the other side.

###### The Int time
In the days of yore, as the User's Guide shows, stackalloc took an `Int`
argument.  Conceptually, this argument is clearly a count.

Yes, C `malloc` & `calloc` use a `size_t` for the argument, but they
also convert (unsigned) and accept integer constants. 
`malloc(5)` compiles & executes quite nicely.  From the SN
code base:
```
platform.c:    char *dest = malloc(9);
```

###### The UInt time

Things being how they are, change came to visit.    The type of the argument
for both `stackalloc` and `alloc` stayed as CSize, but CSize itself changed
from Int to UInt.

This necessitated using `stackalloc[byte](5.toUInt)` all over the place. Note
well that the CSize abstraction is totally broken and worse than useless here.
I have a moldering issue reporting this and the need for a `toCSize`.

There was an explicit decision at the time, by people who should know,
 to not allow implicit conversions from Int to UInt.  

At least UInt is a counting number.

###### The USize time
Time passes and change comes to visit again.  The type underlying
CSize changes to USize. One must now write `stackalloc[byte](5.toUSize)`
all over the place.  By good design, thank you again Shadaj, there is
an implicit from UInt to USize, so all the prior  `stackalloc[byte](5.toUInt)`
continue to work.

###### The Fix time

Imagine a type Count, which is a subset of Int, and holds non-negative Int values.
Imagine an implicit from Count to USize.  All the numbers are known to be non-negative,
so there should be no issues.

Imagine an implicit from Int to Count.   The case of a negative integer must be
handled, probably by Exception, but there should be problem with non-negative arguments.

Now one can write the "natural"  `stackalloc[byte](5)` without running afoul of banned
Int to UInt conversions.

One can collapse the chain of reasoning by eliminating the intermediate and conceptual Count.
It was presented only as a way of developing the reasoning.  It could be implemented if needed
for purity.

#### Alternate fix

This alternative is more straight forward.

An alternate fix, which would avoid a perambulation through the purity maze, would
be to add overloads to both stackalloc and alloc to take an Int argument and check it
for non-negativity.  Yes, this would restrict these overloads from allocating more bytes
at once than 2,147,483,648 on a 32 bit machine or  9,223,372,036,854,775,808
on a 64 bit machine.  One can always use the USize variant for allocations 
larger that that.

**Comment by armanbilge** [comment]
> The fix is to add an implicit from `Int` to (a never actualized) `Count`
> and one from `Count` to `USize`.

I may be misunderstanding your proposed fix, but last I checked implicit conversions cannot be chained/doubled-up. E.g. the compiler will only ever apply at most one implicit conversion, never two in a row e.g. `Int -> Count -> USize`.

**Comment by LeeTibbert** [comment]
Arman, thank you for reading this Issue.

Overall: 
    My motivating goal is to be able to write, at the end of the trail, `stackalloc[Byte](5)` without a lot of
    eyesore faffing around with types which bring no benefit.

My intent with `Count` was that it be a conceptual, never realized, intermediate 'virtual, in the sense of 
virtual particles in physics, popping in and out of existance' type for those concerned that
an implicit directly from `Int` to `USize` might be too great a leap.  

Perhaps I should be describing this as an implicit from 'Int` to `CSize`. There might be
less concern with that.  That is a misdirection, a.k.a joke. In my experience the CSize
abstraction is totally broken/leaky.  But if such a description gets me to my end goal,
I will take it.

A type `Count` could be implemented and an overload  `stackalloc[type](Count)` (so also, `alloc`).
I think that making such an abstraction real adds negative benefit (unless it lets me shout 
"Bingo, I never have to deal with that again").  People have been thinking of integers as counts
as long as they have had fingers, toes, and integers.  

Yes, I should be concerned with Larger Issues and get over this one. 

Epilogue:
Utility brings Int to the front.
Completeness & Parallelism would suggest similar concerns with  other integral types: Byte, Short, & Long.

Imaging the machine capable of executing a `stackalloc` with a Long argument (or even the large 
end of the current USize).

**Comment by LeeTibbert** [comment]
##### Later:   
From lots of attempts to capture it, I believe the annoyance I am seeing about stackalloc[Byte](5) requiring
a (5.toUSize) occurs only in the Scala Native build environment itself.  I have had neither the time nor
skill to unravel the interaction of implicits.

This means that the user facing documentation is correct as is (but it would be a kindness if the 
Contributor's Guide mentioned the quirk and that it, I believe, not seen by people outside
the SN build environment (unless they are examining SN sources as a model!)

A reduction-in-strength of the issue, which means that I will almost certainly never get time
to chase it further.

**Comment by ekrich** [comment]
I ran into this problem as well and used `toULong` but not sure how that would work on 32 bit - a `ULong` should be the same as a `UInt` I would think.

**Comment by WojciechMazur** [comment]
We've added a variant of stackalloc which does accept `Int` literal, tough I belive we can close this issue

**Comment by LeeTibbert** [comment]
@WojciechMazur 

Thank you.  All good things come with time.
